internal/datanode/data_node.go

```diff
@@ -901,6 +900,50 @@ func (node *DataNode) Import(ctx context.Context, req *datapb.ImportTaskRequest)
 	return resp, nil
 }
 
+// AddSegment adds the segment to the current DataNode.
+func (node *DataNode) AddSegment(ctx context.Context, req *datapb.AddSegmentRequest) (*commonpb.Status, error) {
+	log.Info("adding segment to DataNode flow graph",
+		zap.Int64("segment ID", req.GetSegmentId()),
+		zap.Int64("collection ID", req.GetCollectionId()),
+		zap.Int64("partition ID", req.GetPartitionId()),
+		zap.String("channel name", req.GetChannelName()),
+		zap.Int64("# of rows", req.GetRowNum()))
+	// Fetch the flow graph on the given v-channel.
+	ds, ok := node.flowgraphManager.getFlowgraphService(req.GetChannelName())
+	if !ok {
+		log.Error("channel not found in current DataNode",
+			zap.String("channel name", req.GetChannelName()),
+			zap.Int64("node ID", Params.DataNodeCfg.GetNodeID()))
+		return &commonpb.Status{
+			// TODO: Add specific error code.
+			ErrorCode: commonpb.ErrorCode_UnexpectedError,
+		}, nil
+	}
+	// Add the new segment to the replica.
+	if !ds.replica.hasSegment(req.GetSegmentId(), true) {
+		log.Info("add a new segment to replica")
+		err := ds.replica.addNewSegment(req.GetSegmentId(),
+			req.GetCollectionId(),
+			req.GetPartitionId(),
+			req.GetChannelName(),
+			&internalpb.MsgPosition{
+				ChannelName: req.GetChannelName(),
+			},
+			&internalpb.MsgPosition{
+				ChannelName: req.GetChannelName(),
+			})
```

Comment:
Using empty start/end position might be dangerous since it may cause datanode/querynode to consume from the start point of the topic in same cases.